index.js

```diff
@@ -62,15 +61,28 @@ const useNavigate = (options) => {
  * Part 2, Low Carb Router API: Router, Route, Link, Switch
  */
 
-export const Router = (props) => {
-  // this little trick allows to avoid having unnecessary
-  // calls to potentially expensive `buildRouter` method.
+export const Router = ({ hook, matcher, base = "", nested = false, children }) => {
+  const parent = useRouter();
+
+  // nested `<Router />` has the scope of its closest parent router (base path is prepended)
+  // Routers are not nested by default, but this might change in future versions
+  const proto = nested ? parent : defaultRouter;
+
+  // updates the current router with the props passed down to the component
+  const updateRouter = (router) => {
+    router.hook = hook || proto.hook;
+    router.matcher = matcher || proto.matcher;
+    router.base = proto.base + base;
+  };
+
+  // we use `useState` here, but it only catches the first render and never changes.
   // https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily
-  const [value] = useState(() => ({ v: buildRouter(props) }));
+  const [value] = useState(() => ({})); // create the object once...
+  updateRouter(value); // then update it on each render
```

Comment:
@HansBrende here is my first take, no parent yet, but we'll get to that later.
Initially this code was:

```js
// updates the current router with the props passed down to the component
const updateRouter = (router) => {
  router.hook = hook || proto.hook;
  router.matcher = matcher || proto.matcher;
  router.base = proto.base + base;

  return router;
}; 

const [value] = useState(() => updateRouter({})); // create the object once...
useLayoutEffect(() => updateRouter(value)); // then update it on each render
```

However I was receiving `useLayoutEffect` warning when testing it in SSR:
<img width="891" alt="image" src="https://user-images.githubusercontent.com/671276/202009608-116058dd-0daa-4b27-8ca4-98edf0e543c5.png">

Reply:
Any ideas on how we can fix this?

Reply:
@molefrog yes, actually I do. I've seen several big codebases follow a certain pattern to deal with this problem, one moment while I look it up.

Reply:
@molefrog this is how react-redux fixes this problem:

https://github.com/reduxjs/react-redux/blob/master/src/utils/useIsomorphicLayoutEffect.ts

They define a `useIsomorphicLayoutEffect` that resolves to `useLayoutEffect` on the client, and `useEffect` on the server.